### ISSUE TITLE
Function type arrays on annotations causes a bad-sad error

### ISSUE THREAD

**Issue by Ibaqu** [ballerina-platform/ballerina-lang#29795]
**Description:**
Consider the following annotation

```
public type TestConfig record {
    function[] dependsOn = [];
};

public annotation TestConfig Config on function;
```

When the annotation is used, the code compiles however it throws an error with the following stack trace

```
[2021-04-02 14:04:51,698] SEVERE {b7a.log.crash} - error("java.lang.NullPointerException") 
org.ballerinalang.test.runtime.exceptions.BallerinaTestException: error("java.lang.NullPointerException")
	at org.ballerinalang.test.runtime.entity.TesterinaFunction.runOnSchedule(TesterinaFunction.java:122)
	at org.ballerinalang.test.runtime.entity.TesterinaFunction.invoke(TesterinaFunction.java:54)
	at org.ballerinalang.test.runtime.BTestRunner.startSuite(BTestRunner.java:554)
	at org.ballerinalang.test.runtime.BTestRunner.execute(BTestRunner.java:264)
	at org.ballerinalang.test.runtime.BTestRunner.runTest(BTestRunner.java:114)
	at org.ballerinalang.test.runtime.util.TesterinaUtils.executeTests(TesterinaUtils.java:76)
	at org.ballerinalang.test.runtime.Main.startTestSuit(Main.java:116)
	at org.ballerinalang.test.runtime.Main.main(Main.java:97)
Caused by: error("java.lang.NullPointerException")
	at io.ballerina.runtime.api.creators.ErrorCreator.createError(ErrorCreator.java:127)
	at io.ballerina.runtime.internal.ErrorUtils.createInteropError(ErrorUtils.java:78)
	at ballerina.lang$0046value.1_0_0.value.cloneReadOnly(value.bal:53)
	at mohamedaquibzulfikar.Logging.0_1_0.tests.main_test.$annot_func$_0(tests/main_test.bal:3)
	at mohamedaquibzulfikar.Logging.0_1_0.Logging.$gen$$0046$0060testinit$0062(Logging)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.ballerinalang.test.runtime.entity.TesterinaFunction.lambda$runOnSchedule$0(TesterinaFunction.java:109)
	at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(Scheduler.java:571)
	at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:301)
	at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:269)
	at java.base/java.lang.Thread.run(Thread.java:834)
 ```

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

https://github.com/ballerina-platform/ballerina-lang/issues/28358

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

**Comment by MaryamZi** [comment]
Reproducible with 

```ballerina
public function main() {
    function[] arr = [foo];
    function[] & readonly immutableArr = arr.cloneReadOnly();
}

function foo() returns any => 1;
```

Seems to be due to io.ballerina.runtime.internal.types.BFunctionType#hashCode not being updated for the `function` change.

**Comment by MaryamZi** [comment]
Fixed in #29803, merged to master in #29934.
